File: tests/test_color_models.py
"""Tests for RGB and HSV color domain models."""
import pytest
from openrgb_flowers.core.models.color_rgb import ColorRGB
from openrgb_flowers.core.models.color_hsv import ColorHSV


def test_color_rgb_clamping():
    c = ColorRGB(-10, 300, 128)
    assert c.r == 0
    assert c.g == 255
    assert c.b == 128


def test_color_rgb_normalization():
    c = ColorRGB(255, 0, 128)
    norm = c.to_normalized()
    assert norm[0] == 1.0
    assert norm[1] == 0.0
    assert abs(norm[2] - 128 / 255.0) < 1e-4


def test_color_rgb_hex_conversion():
    c = ColorRGB(255, 105, 180)
    assert c.to_hex() == "#FF69B4"
    parsed = ColorRGB.from_hex("#FF69B4")
    assert parsed.r == 255
    assert parsed.g == 105
    assert parsed.b == 180


def test_color_hsv_to_rgb():
    # Pure Red: H=0, S=1, V=1
    hsv_red = ColorHSV(0.0, 1.0, 1.0)
    rgb_red = hsv_red.to_rgb()
    assert rgb_red.r == 255
    assert rgb_red.g == 0
    assert rgb_red.b == 0

    # Pure Green: H=120, S=1, V=1
    hsv_green = ColorHSV(120.0, 1.0, 1.0)
    rgb_green = hsv_green.to_rgb()
    assert rgb_green.r == 0
    assert rgb_green.g == 255
    assert rgb_green.b == 0


def test_color_rgb_to_hsv_roundtrip():
    original = ColorRGB(200, 100, 50)
    hsv = ColorHSV.from_rgb(original)
    recovered = hsv.to_rgb()
    assert abs(recovered.r - original.r) <= 1
    assert abs(recovered.g - original.g) <= 1
    assert abs(recovered.b - original.b) <= 1

File: tests/cpp/test_color_models.cpp
#include "tests/cpp/test_framework.hpp"
#include "openrgb_flowers/core/models/color_rgb.hpp"
#include "openrgb_flowers/core/models/color_hsv.hpp"
#include "openrgb_flowers/core/models/key_coordinate.hpp"
#include "openrgb_flowers/core/models/render_frame.hpp"

using namespace openrgb_flowers::core::models;

TEST_CASE(ColorModels, RGBHexConversion) {
    ColorRGB c1(255, 128, 64);
    EXPECT_EQ(c1.to_hex(), "#FF8040");

    ColorRGB c2 = ColorRGB::from_hex("#FF8040");
    EXPECT_EQ(c2.r, 255);
    EXPECT_EQ(c2.g, 128);
    EXPECT_EQ(c2.b, 64);

    ColorRGB c3 = ColorRGB::from_hex("04080C");
    EXPECT_EQ(c3.r, 4);
    EXPECT_EQ(c3.g, 8);
    EXPECT_EQ(c3.b, 12);
}

TEST_CASE(ColorModels, RGBNormalized) {
    ColorRGB c = ColorRGB::from_normalized(1.0f, 0.5f, 0.0f);
    EXPECT_EQ(c.r, 255);
    EXPECT_NEAR(c.g, 128, 1);
    EXPECT_EQ(c.b, 0);

    auto [rn, gn, bn] = c.to_normalized();
    EXPECT_NEAR(rn, 1.0f, 0.01f);
    EXPECT_NEAR(gn, 0.5f, 0.01f);
    EXPECT_NEAR(bn, 0.0f, 0.01f);
}

TEST_CASE(ColorModels, HSVToRGBRoundtrip) {
    ColorHSV red_hsv(0.0f, 1.0f, 1.0f);
    ColorRGB red_rgb = red_hsv.to_rgb();
    EXPECT_EQ(red_rgb.r, 255);
    EXPECT_EQ(red_rgb.g, 0);
    EXPECT_EQ(red_rgb.b, 0);

    ColorHSV roundtrip = ColorHSV::from_rgb(red_rgb);
    EXPECT_NEAR(roundtrip.h, 0.0f, 0.1f);
    EXPECT_NEAR(roundtrip.s, 1.0f, 0.01f);
    EXPECT_NEAR(roundtrip.v, 1.0f, 0.01f);

    ColorHSV green_hsv(120.0f, 1.0f, 1.0f);
    ColorRGB green_rgb = green_hsv.to_rgb();
    EXPECT_EQ(green_rgb.r, 0);
    EXPECT_EQ(green_rgb.g, 255);
    EXPECT_EQ(green_rgb.b, 0);
}

TEST_CASE(ColorModels, KeyCoordinateClamping) {
    KeyCoordinate k("Esc", 0, -0.5f, 1.5f, 0, 0);
    EXPECT_EQ(k.x, 0.0f);
    EXPECT_EQ(k.y, 1.0f);
}

TEST_CASE(ColorModels, RenderFrameBuffer) {
    std::vector<ColorRGB> buffer = {
        ColorRGB(255, 0, 0),
        ColorRGB(0, 255, 0),
        ColorRGB(0, 0, 255)
    };
    RenderFrame frame(1.5, 42, buffer);
    EXPECT_EQ(frame.led_count, 3);
    EXPECT_EQ(frame.raw_size(), 3 * sizeof(ColorRGB));
    EXPECT_EQ(frame.data()[0].r, 255);
    EXPECT_EQ(frame.data()[1].g, 255);
    EXPECT_EQ(frame.data()[2].b, 255);
}
